sway/security.c

```diff
@@ -0,0 +1,218 @@
+		perm = PRIV_OUTPUT_POWER_MANAGER;
+	else
+		return true;
+
+	uint32_t priv_bit = 1 << perm;
+	uint32_t label_id = security_get_client_label_id(client);
+	uint32_t allowed = security_get_label_access(label_id);
+	if (allowed & priv_bit)
+		return true;
+
+	return false;
+}
+
+static const char* const perm_names[PRIV_LAST] = {
+	"layer_shell",
+	"output_manager",
+	"input_method",
+	"text_input",
+	"foreign_toplevel_manager",
+	"dmabuf_manager",
+	"screencopy_manager",
+	"data_control_manager",
+	"gamma_control_manager",
+	"input_inhibit",
+	"input_keyboard_shortcuts_inhibit",
+	"input_virtual_keyboard",
+	"input_virtual_pointer",
+	"output_power_manager",
+};
+
+enum security_perm security_get_perm_by_name(const char *name)
+{
+	int i;
+	for(i = 0; i < PRIV_LAST; i++) {
+		if (0 == strcmp(name, perm_names[i]))
+			return i;
+	}
+	return PRIV_LAST;
+}
+
+static void security_client_destroy(struct wl_listener *listener, void *client_v)
+{
+	const struct wl_client *client = client_v;
+	if (!security_client_list)
+		return;
+
+	for (int i = 0; i < security_client_list->length; ++i) {
+		struct security_client *item = security_client_list->items[i];
+		if (client != item->client)
+			continue;
+		list_del(security_client_list, i);
+		free(item);
+		return;
+	}
+}
+
+uint32_t security_get_client_label_id(const struct wl_client *client)
+{
+	if (client && security_client_list) {
+		for (int i = 0; i < security_client_list->length; ++i) {
+			struct security_client *item = security_client_list->items[i];
+			if (client == item->client)
+				return item->label_id;
+		}
+	}
+	return 0;
+}
+
+void security_set_client_label_id(struct wl_client *client, uint32_t label_id) {
+	struct security_client *found = NULL;
+	if (!security_client_list)
+		security_client_list = create_list();
+	for (int i = 0; i < security_client_list->length; ++i) {
+		struct security_client *item = security_client_list->items[i];
+		if (client == item->client) {
+			found = item;
+			break;
+		}
+	}
+	if (!found) {
+		found = malloc(sizeof(*found));
+		found->client = client;
+		found->on_destroy.notify = security_client_destroy;
+		list_add(security_client_list, found);
+		wl_client_add_destroy_listener(client, &found->on_destroy);
+	}
+
+	found->label_id = label_id;
+}
+
+uint32_t security_get_label_access(uint32_t label_id) {
+	if (!security_labels)
+		return ~0;
+	if (label_id > security_label_len)
+		label_id = 0;
+	return security_labels[label_id].allowed;
+}
+
+void security_set_label_access(uint32_t label_id, uint32_t allowed) {
+	if (label_id < security_label_len)
+		security_labels[label_id].allowed = allowed;
+}
+
+uint32_t security_get_label_id(const char* name)
+{
+	size_t i;
+	if (!security_labels) {
+		security_label_capacity = 8;
+		security_label_len = 1;
+		security_labels = calloc(sizeof(security_labels[0]), security_label_capacity);
+		security_labels[0].allowed = ~0;
+		security_labels[0].name = "default";
```

Comment:
Do we really need "default"? I don't think we implicitly associate everything that is not explicitly labeled with "default" label.

Reply:
We need some way to set the permissions of clients that don't get a label from somewhere else. This lets you do things like disable protocols you don't want to use at all (as discussed in #5583).